text/0000-alternate-registries.md

```diff
@@ -0,0 +1,120 @@
+open source projects using Rust, however for closed source code this is problematic. Currently
+the only real option is to use Git repositories to specify the packages, but that means that
+all of the nice versioning and discoverability that cargo and crates.io provides is lost. I
+would like to change this so that it is possible to have a local crates.io server which private
+crates can be pushed to, plus still be able to make use of the public crates.io server as well.
+
+# Detailed design
+[design]: #detailed-design
+
+There are a number of different areas which will likely need to be tackled in order to fully
+support a local crates.io server in an enterprise. Below are some key areas:
+
+* Add support for alternate crates.io registries (this RFC)
+* Provide support for caching crates used on a crates.io proxy
+* Add support to crates.io to allow authentication with other OAuth providers than github
+* Support for private storage of crates on crates.io server rather than publicly available on S3
+
+Rather than trying to get everything agreed in a single RFC I would like for these to be
+dealt with as separate proposals and use this RFC as a stepping stone to be able to
+meaningfully start to use private registries.
+
+See https://github.com/rust-lang/cargo/pull/4036 for the code changes which this proposal was based on.
+
+## Cargo.toml config changes
+The following changes for Cargo.toml are proposed as part of this RFC:
+* Add an optional 'registry' field to be specified as part of the package dependencies
+* Add an optional 'registry' field to be specified in the package definition
+
+Both of the registry entries take a value of a crates.io git repository. If one is not
+provided then the default crates.io repository is assumed, this ensures that it is
+back compatible with all current crates.
+
+Below is an example of the change that we are proposing to make:
+
+```
+[package]
+name = "registry-test"
+version = "0.1.0"
+authors = ["Christopher Swindle <christopher.swindle@metaswitch.com>"]
+registry = "https://github.com/my_awesome_fork/crates.io-index"
+
+[dependencies]
+libc = { version = "*", registry = "https://github.com/my_awesome_fork/crates.io-index" }
+serde_json = "1.0.1"
+```
+
+## Changes for alternative registries for dependencies
+This boils down to a very simple change, where we previously setup the crate source for the
+crates.io registry, we now just need to check if a registry is provided, if it has the crate
+source is created using the registry URL, otherwise the crates.io server is used.
+
+## Blocking requests to push to a registry
+There are two parts to this, the first is a change to Cargo which checks if the registry
+provided in the registry matches the host for the publish, if it does not it gets rejected.
+The second part is a change to crates.io which will just reject the request to publish the
+crate if the configured repository on the crates.io server does not match the registry
```

Comment:
It needs to know the git repository and compare that, which it already has, so I don't think that this is an issue.